src/bp/ml/toolkit.ts

```diff
@@ -75,23 +76,45 @@ if (cluster.isWorker) {
   }
 }
 
-if (cluster.isMaster) {
-  // cancel svm training once implemented in node binding
-  // registerMsgHandler('cancel', async (msg: Message, worder) => {
-  // })
-
-  registerMsgHandler('train', async (msg: Message, worker) => {
-    const sendToWorker = (msg: Message) => worker.isConnected() && worker.send(msg)
+if (cluster.isWorker) {
+  if (process.env.WORKER_TYPE === <WorkerType>'WEB_WORKER') {
+    overloadTrainers()
+  }
+  if (process.env.WORKER_TYPE === <WorkerType>'ML_WORKER') {
+    async function messageHandler(msg: Message) {
+      if (msg.type === 'train') {
+        const svm = new SVMTrainer()
+        try {
+          let progressCalls = 0
+          const result = await svm.train(msg.payload.points, msg.payload.options, progress => {
+            if (++progressCalls % 5 === 0) {
+              process.send!({ type: 'progress', id: msg.id, payload: { progress } })
+            }
+          })
 
-    const svm = new SVMTrainer()
-    try {
-      const result = await svm.train(msg.payload.points, msg.payload.options, progress =>
-        sendToWorker({ type: 'progress', id: msg.id, payload: { progress } })
-      )
-      sendToWorker({ type: 'done', id: msg.id, payload: { result } })
-    } catch (error) {
-      sendToWorker({ type: 'error', id: msg.id, payload: { error } })
+          process.send!({ type: 'done', id: msg.id, payload: { result } })
+        } catch (error) {
+          process.send!({ type: 'error', id: msg.id, payload: { error } })
+        }
+      }
     }
+
+    process.on('message', messageHandler)
+  }
+}
+
+if (cluster.isMaster) {
+  function sendToWebWorker(msg: Message) {
+    const webWorker = cluster.workers[process.WEB_WORKER]
+    webWorker?.isConnected() && webWorker.send(msg)
+  }
+
+  registerMsgHandler('done', msg => sendToWebWorker(msg))
+  registerMsgHandler('progress', msg => sendToWebWorker(msg))
+  registerMsgHandler('error', msg => sendToWebWorker(msg))
+  registerMsgHandler('train', async (msg: Message) => {
+    const workerID = _.sample(process.ML_WORKERS)
```

Comment:
Should we keep trace of idle/busy workers to distribute the load ? Sample is a bit too random, especially with a low number of workers

Reply:
I thought of it for sure but didnt want to make it over complexe fir a 1st implementation.